docker-images/kafka-based/kafka/scripts/kafka_liveness.sh

```diff
@@ -1,11 +1,19 @@
 #!/usr/bin/env bash
 set -e
 
+# Test ZK-based broker and KRaft broker liveness
+# For ZK-based broker we expect that either the broker is ready and listening on 9091 (replication port) or it has a ZK session
+# For KRaft broker if it is not ready, we mark it as alive, otherwise we check it is listening on 9091 (replication)
 if [ -f /var/opt/kafka/kafka-ready ] ; then
   rm -f /var/opt/kafka/zk-connected 2&> /dev/null
   # Test listening on replication port 9091
   netstat -lnt | grep -Eq 'tcp6?[[:space:]]+[0-9]+[[:space:]]+[0-9]+[[:space:]]+[^ ]+:9091.*LISTEN[[:space:]]*'
-else
-  # Not yet ready, so test ZK connected state
+elif [ "$STRIMZI_KRAFT_ENABLED" != "true" ]; then
+  # Not yet ready, so if in ZK mode test ZK connected state
   test -f /var/opt/kafka/zk-connected
 fi
+
+if [ "$STRIMZI_KRAFT_ENABLED" = "true" ]; then
+  # Test KRaft controller liveness
+  . ./kafka_controller_liveness_readiness.sh
+fi
```

Comment:
This seems a bit weird TBH. 

In Zoo mode ... 
* If Kafka is ready, we test port 9091. 
* If it isn't ready, we test if it connected to ZooKeeper

In KRaft mode:
* If Kafka is ready, we test port 9091 and 9090
* If Kafka is not ready, we test only 9090

Does Kafka actually treat these listeners differently? It did not used to in the past and IMHO the original probe is simply using 9091 because that is what we had. 9090 was added only later. does it really make sense to test both ports?

Reply:
I think combining the broker and controller checks in the above makes the comparison seem weird. A fairer comparison is:

In Zoo mode:
 - If Kafka is ready, we test port 9091.
 - If it isn't ready, we test if it connected to ZooKeeper

In KRaft mode:
For brokers:
 - If Kafka is ready, we test port 9091.
 - If it isn't ready we just mark it as alive.
For controllers:
 - We test for port 9090

So my understanding is the 9090 port is specific to KRaft controllers and is not relevant to KRaft brokers. Does that make more sense? I tried to explain it clearly in the PR description.

Reply:
Yeah, but right now this is for a combined node. Do we really want to test both port 9090 and 9091 at the same time on combined nodes?

Reply:
> If it isn't ready we just mark it as alive

This sounds weird to me but I have not deeper knowledge of brokers in KRaft mode but how can say broker is alive but not ready?

> Yeah, but right now this is for a combined node. Do we really want to test both port 9090 and 9091 at the same time on combined nodes?

Can it happen that a combined node is listening on the replication port 9091 but not on 9090 so it cannot be considered working properly?

Reply:
The ports are used for two different things, so I think for a combined mode to work correctly it would need to be listening on both. I don't know enough about how KRaft mode works under the covers to know for sure whether a combined node could get into a state where it has one listener working and one not.

I did a very basic test where I tried starting a combined node with the replication port already in use, and then the controller port already in use and in both cases the combined node fully shuts down. It looks like it attempts to start the controller first, then if that is successful it starts the broker. If the broker then hits an error that causes a shutdown it shuts down first and then the controller shuts down.

If we were to only check one port I guess the replication port would be the one to check, since it appears to listen on that port second.

@tombentley @showuon what do you think?

Reply:
>> If it isn't ready we just mark it as alive
>
> This sounds weird to me but I have not deeper knowledge of brokers in KRaft mode but how can say broker is alive but not ready?

I don't think it's really about KRaft mode, it's more about Kubernetes views on status. My shorthand description:  
- `liveness` is the pod doing something useful. 
- `readiness` can it accept traffic. 

So as an example we would want to mark a Kafka broker as alive in `broker_state == 2` a.k.a LogRecovery but we wouldn't mark it as ready to accept traffic as the listeners are not available yet.

Apologies if this is all to basic I just want to make sure we are all on the same page. 

In combined mode we should probably mark the pod as alive if the controller port is listening and ready when the listener port is available. 

Is port availability enough to mark a controller as ready? Should we be waiting for it become a member of the quorum? Can we tell when it is?